### ISSUE TITLE
Incorrect BigDecimal Format! output

### ISSUE THREAD

**Issue by DBoehm-Superstate** [akubera/bigdecimal-rs#140]
BigDecimal does not format correctly for some values.

```
let result:BigDecimal = BigDecimal::from(999999991) / 1e6;
assert_eq!(format!("{:.2}", result), "1000.00");
```

I would expect this assert to succeed, but format!("{:.2}", result) yields "100.00". I could at least believe "999.99" but "100.00" seems like a bug.

Interestingly both of the below give the expected answer.
```
assert_eq!(format!("{:.2}", result.to_f64().unwrap()), "1000.00");
assert_eq!(format!("{:.2}", result.round(2)), "1000.00");
```

**Comment by DBoehm-Superstate** [comment]
```       mod dec_9d99 {
            use super::*;

            fn test_input() -> BigDecimal {
                "9.99".parse().unwrap()
            }

            impl_case!(fmt_default:      "{}" => "9.99");
            impl_case!(fmt_d0:        "{:.0}" => "10");
            impl_case!(fmt_d1:        "{:.1}" => "10.0");
            impl_case!(fmt_d2:        "{:.2}" => "9.99");
            impl_case!(fmt_d3:        "{:.3}" => "9.990");
        }```

**Comment by akubera** [comment]
It’s rounding to the nearest 2 decimal digits, so it will round up to 100.00, rather than truncate.

If you want a truncated decimal to 2 places, you can use `result.with_scale(2)`

**Comment by akubera** [comment]
Maybe I misunderstood. It’s saying 100.00 but should be 1000.00? yeah that’s a bug, but might be fixed in a recent pull where round was reimplemented, can you check if `trunk` works?

**Comment by akubera** [comment]
I've cleaned and reorganized the `impl_fmt.rs` file in my local repo, fixing this issue. I'll have that up and released soon. 

This boils down not to a off-by-one error like I expected, but using the same function for formatting scientific notation and "full" decimals. When rounding caused overflow it incremented the exponent `9e-1 -> 1e0` but the "full" formatting code ignored the exponent.

I've refactored the relevant rounding code to a new function which returns the exponent of the rounded digits, and both functions will handle exponent changes appropriately.

**Comment by akubera** [comment]
174 unit tests and [100% code coverage](https://app.codecov.io/gh/akubera/bigdecimal-rs/commit/4c5ee5af4168a5aaac1493da6947dd649192c801/blob/src/impl_fmt.rs) in `impl_fmt.rs` wasn't enough. 😮‍💨 

I had `0.9`, but not `0.99`, and that took the wrong branch.

**Comment by akubera** [comment]
Fixed in v0.4.6.

Thanks for reporting this.

**Comment by DBoehm-Superstate** [comment]
Sorry I didn't check back in quickly enough to respond to your question.  Thanks for fixing this!

**Comment by akubera** [comment]
Oh that's fine. I was on my way to dinner when I read (and subsequently re-read) the issue, and couldn't properly check it. I was just hoping it had already been fixed in trunk, but it wasn't.

I added some property tests to see if that would have caught the overflow bug, and it actually did _not._ So thanks again for letting me know.

\<aside>

property tests did reveal that if `n = 1.0437336860498237e-47` then `{n:.0}` will be formatted as `1e-47`, and I think that should simply be `0`. So decimal formatting is still not complete.
